### PR TITLE
Mycus mutation category now works again, among other things

### LINKED ISSUE #58747: Mycus mutation broken
### Describe the bug

It is not possible to get mycus traitsfeeder, identity, etc.), except for the just "mycus" (. Sleep, eating an incredible amount of mycus fruits and wine do not help either.

### Steps to reproduce

1. Eat marloss berries and seeds.
2. Go to fungus tower and let it give you "marloss gate" trait.
3. Eat mycus fruit, get "mycus" trait
4. Eat more mycus fruits, drink mycus wine and sleep, to mutate further.
5. Wait eternity and die without getting any other trait.

### Expected behavior

After eating mycus fruits and sleeping for a week character should get some new traits, like mycus identity, feeder, mycus flesh, defender and other.

### Screenshots

_No response_

### Versions and configuration

- OS: Win 10
- Game version: experimental 2022-06-26-1548
- Graphic version: tiles
- Ingame language: rus
- Mods loaded: [
  "dda",
  "no_npc_food",
  "package_bionic_professions",
  "StatsThroughSkills",
  "stats_through_kills"
]

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Ilysen**
#### Summary
Bugfixes "The Mycus now accepts local guides properly"

#### Purpose of change

Fixes #58747

Also fixes the related issue of the rat king monster not causing mutations correctly.

#### Describe the solution

This bug is twofold:
1. Dreams did not display without vitamins, so you wouldn't ever get dream messages as a Mycus character;
2. The function to mutate towards a category was broken. Despite having an argument to let vitamin requirements be bypassed, this argument wasn't passed to subsequent logic, so vitamins were needed anyway. The Mycus line, which lacks a vitamin type due to its special handling, would thus always fail to mutate.

The fixes are as follows:
1. If the character has crossed a threshold and that threshold has a category entry, always show dreams for that category. Right now this only affects Mycus; other threshold mutations lack category entries, and I'm not sure if that's intentional.
2. The `use_vitamins` argument is now passed to subsequent logic, ensuring that vitamin requirements are correctly waived in those instances. This fixes both the Mycus mutation tree as well as involuntary mutations caused by the `rat king` monster.

I also made these mutations true random, as they are separated from the instability system.

#### Describe alternatives you've considered

There are some idiosyncrasies with the mutation system that I'm not sure about; for instance, I'm not sure if it's intended that threshold mutations have no category listed, and if adding one for the Mycus trait will break things. Ultimately it proved to be the simplest way to fix the issue, but I'm very much open to hear if there's a reason for that.

#### Testing

* I used the map editor to make a little jail for the rat king and spawned him inside there, then waited for a little bit to make sure that mutations were happening. I used debug messages early on to make sure that all mutations were being targeted, and not just positive ones.
* I got the Mycus threshold, both through debug and the "natural" way, and tried both eating Mycus fruit (which has a 1 in 3 chance to give you Mycus mutations if you don't have `Mycus Feeder`) and sleeping. I was able to mutate in the category again, and dreams were appearing again as they once had.
* I gave myself medical primer and mutagen for a little while, just to make sure that regular mutating didn't seem busted.

#### Additional context

Immense thanks goes to @dseguin for helping me with the debugging and making of this.

image
image

### CODE REVIEW

**Comment by dseguin** [comment]
You can try these changes to get "use_vitamins" to propagate:

<details>
<summary>(diff)</summary>

```diff
diff --git a/src/character.cpp b/src/character.cpp
index 242b520092..0a0fd1a2e7 100644
--- a/src/character.cpp
+++ b/src/character.cpp
@@ -7791,7 +7791,7 @@ bool Character::crossed_threshold() const
 mutation_category_id Character::get_threshold_category() const
 {
     for( const trait_id &mut : get_mutations() ) {
-        if( mut->threshold ) {
+        if( mut->threshold && !mut->category.empty() ) {
             return mut->category[0];
         }
     }
diff --git a/src/character.h b/src/character.h
index 7ac5887d07..138ddf7169 100644
--- a/src/character.h
+++ b/src/character.h
@@ -1374,10 +1374,10 @@ class Character : public Creature, public visitable
         void mutate_category( const mutation_category_id &mut_cat );
         /** Mutates toward one of the given mutations, upgrading or removing conflicts if necessary */
         bool mutate_towards( std::vector<trait_id> muts, const mutation_category_id &mut_cat,
-                             int num_tries = INT_MAX );
+                             int num_tries = INT_MAX, bool use_vitamins = true );
         /** Mutates toward the entered mutation, upgrading or removing conflicts if necessary */
         bool mutate_towards( const trait_id &mut, const mutation_category_id &mut_cat,
-                             const mutation_variant *chosen_var = nullptr );
+                             const mutation_variant *chosen_var = nullptr, bool use_vitamins = true );
         bool mutate_towards( const trait_id &mut, const mutation_variant *chosen_var = nullptr );
         /** Removes a mutation, downgrading to the previous level if possible */
         void remove_mutation( const trait_id &mut, bool silent = false );
diff --git a/src/mutation.cpp b/src/mutation.cpp
index b0497453ce..3c7c2f4150 100644
--- a/src/mutation.cpp
+++ b/src/mutation.cpp
@@ -1064,7 +1064,7 @@ void Character::mutate( const int &true_random_chance, const bool use_vitamins )
                 size_t roll = rng( 0, upgrades.size() + 4 );
                 if( roll < upgrades.size() ) {
                     // We got a valid upgrade index, so use it and return.
-                    mutate_towards( upgrades[roll], cat );
+                    mutate_towards( upgrades[roll], cat, nullptr, use_vitamins );
                     return;
                 }
             }
@@ -1086,7 +1086,7 @@ void Character::mutate( const int &true_random_chance, const bool use_vitamins )
         if( !dummies.empty() ) {
             std::shuffle( dummies.begin(), dummies.end(), rng_get_engine() );
             for( trait_id &tid : dummies ) {
-                if( has_conflicting_trait( tid ) && mutate_towards( tid, cat ) ) {
+                if( has_conflicting_trait( tid ) && mutate_towards( tid, cat, nullptr, use_vitamins ) ) {
                     add_msg_if_player( m_mixed, mutation_category_trait::get_category( cat ).mutagen_message() );
                     return;
                 }
@@ -1107,7 +1107,7 @@ void Character::mutate( const int &true_random_chance, const bool use_vitamins )
             if( mut_vit != vitamin_id::NULL_ID() && vitamin_get( mut_vit ) >= 2200 ) {
                 test_crossing_threshold( cat );
             }
-            if( mutate_towards( valid, cat, 2 ) ) {
+            if( mutate_towards( valid, cat, 2, use_vitamins ) ) {
                 add_msg_if_player( m_mixed, mutation_category_trait::get_category( cat ).mutagen_message() );
             }
             return;
@@ -1150,12 +1150,12 @@ void Character::mutate_category( const mutation_category_id &cat, const bool use
         }
     }
 
-    mutate_towards( valid, cat, 2 );
+    mutate_towards( valid, cat, 2, use_vitamins );
 }
 
 void Character::mutate_category( const mutation_category_id &cat )
 {
-    mutate_category( cat, false );
+    mutate_category( cat, !mutation_category_trait::get_category( cat ).vitamin.is_null() );
 }
 
 static std::vector<trait_id> get_all_mutation_prereqs( const trait_id &id )
@@ -1175,12 +1175,12 @@ static std::vector<trait_id> get_all_mutation_prereqs( const trait_id &id )
 }
 
 bool Character::mutate_towards( std::vector<trait_id> muts, const mutation_category_id &mut_cat,
-                                int num_tries )
+                                int num_tries, bool use_vitamins )
 {
     while( !muts.empty() && num_tries > 0 ) {
         int i = rng( 0, muts.size() - 1 );
 
-        if( mutate_towards( muts[i], mut_cat ) ) {
+        if( mutate_towards( muts[i], mut_cat, nullptr, use_vitamins ) ) {
             return true;
         }
 
@@ -1192,7 +1192,7 @@ bool Character::mutate_towards( std::vector<trait_id> muts, const mutation_categ
 }
 
 bool Character::mutate_towards( const trait_id &mut, const mutation_category_id &mut_cat,
-                                const mutation_variant *chosen_var )
+                                const mutation_variant *chosen_var, bool use_vitamins )
 {
     if( has_child_flag( mut ) ) {
         remove_child_flag( mut );
@@ -1297,9 +1297,9 @@ bool Character::mutate_towards( const trait_id &mut, const mutation_category_id
 
     if( !c_has_both_prereqs && ( !prereqs1.empty() || !prereqs2.empty() ) ) {
         if( !c_has_prereq1 && !prereqs1.empty() ) {
-            return mutate_towards( prereqs1, mut_cat );
+            return mutate_towards( prereqs1, mut_cat, INT_MAX, use_vitamins );
         } else if( !c_has_prereq2 && !prereqs2.empty() ) {
-            return mutate_towards( prereqs2, mut_cat );
+            return mutate_towards( prereqs2, mut_cat, INT_MAX, use_vitamins );
         }
     }
 
diff --git a/src/mutation_data.cpp b/src/mutation_data.cpp
index 7aefecc4ed..d8a66f9511 100644
--- a/src/mutation_data.cpp
+++ b/src/mutation_data.cpp
@@ -109,7 +109,7 @@ void mutation_category_trait::load( const JsonObject &jsobj )
     static_cast<void>( translate_marker_context( "memorial_female", "Crossed a threshold" ) );
     optional( jsobj, false, "memorial_message", new_category.raw_memorial_message,
               text_style_check_reader(), "Crossed a threshold" );
-    new_category.vitamin = vitamin_id( jsobj.get_string( "vitamin", "" ) );
+    new_category.vitamin = vitamin_id( jsobj.get_string( "vitamin", "null" ) );
 
     mutation_category_traits[new_category.id] = new_category;
 }
```

</details>

Tested briefly and it seems to work

Screenshot from 2022-10-18 02-06-09

**Comment by Ilysen** [comment]
no idea why the test is failing, so I'm open to hear out help. :p

File: src/character.h [hunk 1]
```diff
@@ -1370,14 +1370,15 @@ class Character : public Creature, public visitable
...
+                              bool true_random = false );
         void mutate_category( const mutation_category_id &mut_cat );
         /** Mutates toward one of the given mutations, upgrading or removing conflicts if necessary */
         bool mutate_towards( std::vector<trait_id> muts, const mutation_category_id &mut_cat,
-                             int num_tries = INT_MAX );
+                             int num_tries = INT_MAX, bool use_vitamins = true );
         /** Mutates toward the entered mutation, upgrading or removing conflicts if necessary */
         bool mutate_towards( const trait_id &mut, const mutation_category_id &mut_cat,
-                             const mutation_variant *chosen_var = nullptr );
+                             const mutation_variant *chosen_var = nullptr, const bool use_vitamins = true );
```
Comment (dseguin) [inline src/character.h:1381]:
```suggestion
                             const mutation_variant *chosen_var = nullptr, bool use_vitamins = true );
```
Clang-tidy is complaining about an unnecessary const here (basic data types are passed by value so making it const doesn't really make a difference)

File: src/mutation.cpp [hunk 2]
```diff
@@ -1192,7 +1193,7 @@ bool Character::mutate_towards( std::vector<trait_id> muts, const mutation_categ
 }
 
 bool Character::mutate_towards( const trait_id &mut, const mutation_category_id &mut_cat,
-                                const mutation_variant *chosen_var )
+                                const mutation_variant *chosen_var, const bool use_vitamins )
```
Comment (dseguin) [inline src/mutation.cpp:1196]:
```suggestion
                                const mutation_variant *chosen_var, bool use_vitamins )
```

**Review by dseguin — APPROVED** [review]

**Pull request merged**